SnoopCompileBasics.ipynb

```diff
@@ -0,0 +1,717 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Presenter: restart kernel and clear output!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "using Pkg\n",
+    "Pkg.activate(\".\")\n",
+    "Pkg.precompile()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "# SnoopCompile basics\n",
+    "\n",
+    "## Part 3 of \"Package development: improving engineering quality & latency,\" JuliaCon 2021\n",
+    "\n",
+    "Tim Holy\n",
+    "\n",
+    "- Collecting \"snoop\" data\n",
+    "- Graphical tools\n",
+    "  + flamegraphs (e.g., profiling)\n",
+    "  + profile-guided despecialization\n",
+    "- `precompile` statement generation\n",
+    "- More reasons to want inferrability"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "# Comparing JET and SnoopCompile\n",
+    "\n",
+    "JET (see previous presentation) has both similarities and differences with [SnoopCompile](https://github.com/timholy/SnoopCompile.jl)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "Most obviously, SnoopCompile focuses on *latency* (shortening \"time to first plot\") whereas JET focuses on *correctness*."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "However, there are deep commonalities. A key reason is that both packages \"spy\" on type-inference to gather data about code."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "But even here they have differences."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "source": [
+    "```julia\n",
+    "sum2(list) = [sum(list[1]), sum(list[2])]\n",
+    "\n",
+    "data = Any[[1,2,3], Any[1,2,3]]\n",
+    "\n",
+    "sum2(data)\n",
+    "\n",
+    "# sum(Any[1,2,3])\n",
```

Comment:
```suggestion
    "# sum2(Any[[1,2,3], []])\n",
```
maybe ?

Reply:
It's there to show the line that JET does pick up on.